File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/RxStorage.java
package com.example.alexeyglushkov.cachemanager;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.reactivex.Completable;
import io.reactivex.Maybe;
import io.reactivex.Single;

public interface RxStorage {
    Completable put(@NonNull String key, @NonNull Object value, @Nullable StorageMetadata metadata);

    Maybe<Object> getValue(@NonNull String key);

    Maybe<StorageMetadata> getMetadata(@NonNull String key);
    Completable remove(@NonNull String key);

    Maybe<StorageEntry> getEntry(@NonNull String key);
    Single<List<StorageEntry>> getEntries();
    Completable removeAll();
}

File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/RxStorage.kt
package com.example.alexeyglushkov.cachemanager

import io.reactivex.Completable
import io.reactivex.Maybe
import io.reactivex.Single

interface RxStorage {
    fun put(key: String, value: Any, metadata: StorageMetadata): Completable
    fun getValue(key: String): Maybe<Any>
    fun getMetadata(key: String): Maybe<StorageMetadata>
    fun remove(key: String): Completable
    fun getEntry(key: String): Maybe<StorageEntry>
    fun getEntries(): Single<List<StorageEntry>>
    fun removeAll(): Completable
}